components/chapters/LoginModal.tsx

```diff
@@ -0,0 +1,196 @@
+  const [user, setUser] = useState<any>({});
+  const [copied, setCopied] = useState(false);
+
+  function loadProgress() {
+    const privateKey = Secp256k1.uint256(Buffer.from(userPrivateKey, 'hex'), 16)
+    const publicKey = Secp256k1.generatePublicKeyFromPrivateKeyData(privateKey);
+
+    let user: any = window.localStorage.getItem('user')
+    if (user) {
+      user = JSON.parse(user)
+
+      if (user.publicKey.x === publicKey.x && user.publicKey.y === publicKey.y) {
+        window.localStorage.setItem('loggedIn', "true");
+        setUser(user);
+        setLoaded(true);
+      } else {
+        alert('Invalid private key')
+      }
+    }
+  }
+
+  function checkValidKey() {
+    if (userPrivateKey.length !== 64) return false;
+
+    const user = window.localStorage.getItem('user');
+    if (user) {
+      const userObj = JSON.parse(user);
+      const privateKey = Secp256k1.uint256(Buffer.from(userPrivateKey, 'hex'), 16)
+      const publicKey = Secp256k1.generatePublicKeyFromPrivateKeyData(privateKey);
+
+      if (userObj.publicKey.x === publicKey.x && userObj.publicKey.y === publicKey.y) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  function clearProgress() {
+    window.localStorage.removeItem('user');
+    window.localStorage.removeItem('loggedIn');
+
+    props.onClose();
+  }
+
+  function copy() {
+    navigator.clipboard.writeText(userPrivateKey)
+    setCopied(true)
+
+    setTimeout(() => {
+      setCopied(false)
+    }, 2000)
+  }
+
+  function signOut() {
+    setLoaded(false);
+    props.onLogout();
+    props.onClose();
+  }
+
+  return (
+    <Modal
+      isOpen={props.open}
+      className="top-1/2 left-1/2 outline-none p-5 transform -translate-x-1/2 -translate-y-1/2 absolute w-[32rem] rounded-lg shadow-lg bg-[#32455D] text-white"
+      contentLabel="Login Modal"
+    >
+      <div className="float-right flex justify-end">
+        <button onClick={props.onClose} aria-label="Close">
+          <svg
+            xmlns="http://www.w3.org/2000/svg"
+            className="h-6 w-6"
+            fill="none"
+            viewBox="0 0 24 24"
+            stroke="currentColor"
+          >
+            <path
+              strokeLinecap="round"
+              strokeLinejoin="round"
+              strokeWidth={2}
+              d="M6 18L18 6M6 6l12 12"
+            />
+          </svg>
+        </button>
+      </div>
+
+      <div className="p-4">
+      {!loaded && !props.signedIn && (
+        <>
+          <img src="/assets/avatars/1.png" width={80} height={80} />
+          <h2 className="mb-4 mt-5 text-3xl font-bold">Load your progress</h2>
+          <p className="mb-5 text-xl">
+            Enter your personal code below to restore your chapter and challenge progress, as well as your avatar. <a href="#" className='text-white underline'>No code yet?</a>
+          </p>
+
+          <div className="flex flex-col justify-between">
+            <div className="flex w-full">
```

Comment:
Why do you need these wrapper div tags around every UI element?

Reply:
I like to keep UI elements separated by divs but I can change it if you want.

Reply:
Why?

Reply:
Sometimes it's easier to wrap form components with a div to add styles in css so it became a habit.